### ISSUE TITLE
centralManager unable to be used after DFU

### ISSUE THREAD

**Issue by qvivr-cris** [nordicsemi/IOS-DFU-Library#29]
We are allocating our CBCentralManager as so:

myCentralManager = [[CBCentralManager alloc] initWithDelegate:self queue:nil options:@{CBCentralManagerOptionRestoreIdentifierKey : @"MyBTCentral"}];

We perform a DFU and we attempt to reconnect to our device via the device UUID
eg: 
  NSArray *peripheralArray = [myCentralManager retrievePeripheralsWithIdentifiers:@[nsUUID]];
    if([peripheralArray count] > 0)
    {
         [myCentralManager connectPeripheral: peripheralArray[0] options:nil];  
    }

At this point, centralManager centralManagerDidUpdateState delegate is called and the centralManager state is  CBManagerStateUnsupported 

Is there a way for us to reset the  centralManager ?

**Comment by qvivr-cris** [comment]
(my apologies if this is posted in incorrect forum or location)

**Comment by mostafaberg** [comment]
The device's UUID might change after being flashed depending on the new firmware, are you sure that the device isn't broadcasting a different UUID at that time ?

also, are you using `Obj-c` ?

**Comment by qvivr-cris** [comment]
Hi, thanks for the response. 
Yes I'm using objective C. 
I should add that killing the App, and relaunching it, in which it re-initialize myCentralManager  it does connect to the device fine.  
I've tried to reinit the myCentralManager without relaunching and again I get CBManagerStateUnsupported.   I could try it without the CBCentralManagerOptionRestoreIdentifierKey to see if that is a factor, but it only would be a test, as we generally do need the use of the CBCentralManagerOptionRestoreIdentifierKey.

**Comment by mostafaberg** [comment]
Yeah that was my next tip, can you try to avoid using the options and recreate a new CentralManager and see if it works ? not the solution, just testing to troubleshoot :)

also, can you verify the peripheral has the same UUID you're trying to connect to ? you can use our [nRFConnect](https://itunes.apple.com/no/app/nrf-connect/id1054362403?mt=8) app on the app store to do that

**Comment by qvivr-cris** [comment]
Yes, its the same UUID, as we use the UUID when we relaunch the App to connect. 
I will try reinit the myCentralManager without the  CBCentralManagerOptionRestoreIdentifierKey and see what results I get.

**Comment by mostafaberg** [comment]
Ok so UUID is not your issue for sure, try without the restoration setting and see if it works out

**Comment by xiejiajian** [comment]
I am looking forward to your solution,cause I have the familiar problem, even if i don't use  CBCentralManagerOptionRestoreIdentifierKey.
And i have an additional issue, when i use obj-c to create a 'dfuViewController' just like your dfu Sample, it seems not working as i tapped the 'resume'
Hope there is a guide ,thanks!

```
UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"NO" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
            [weakSelf logWith:LogLevelVerbose message:@"Abort cancelled"];
            if (weakSelf.dfuController.paused) {
                [weakSelf.dfuController resume];
            }
        }];
```

**Comment by mostafaberg** [comment]
@qvivr-cris Do you have any updates yet ?
@xiejiajian please open a new issue to avoid confusion and i'll help you there!

**Comment by calvingit** [comment]
After DFU, centralManager's delegate is taken over by DFU Library. So, the original  centralManager can't work with CBCentralManagerDelegate methods. You should reset the delegate of centralManager to belong  to your class.

**Comment by qvivr-cris** [comment]
I believe the issue is that we were using CBCentralManagerOptionRestoreIdentifierKey, and initializing a 2nd time, which would fail. A 2nd initialization without a key worked fine, and we were able to reconnect via the device via iOS API of 

NSArray *peripheralArray = [myCentralManager retrievePeripheralsWithIdentifiers:@[nsUUID]];
[myCentralManager connectPeripheral: peripheralArray[0] options:nil];

**Comment by casey-juul** [comment]
To @calvingit 's comment, we re-establish the central manager delegate to our own class when the DFU process is done or interrupted. The following code seems to work in all scenarios except one edge-case interruption. This basically listens to when the Nordic SDK is done with the delegate by setting it to nil:

```
- (void)startUpdate {
    [myCentralManager addObserver:self forKeyPath:@"delegate" options:NSKeyValueObservingOptionOld | NSKeyValueObservingOptionNew context:nil];

    // start DFU.
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSString *,id> *)change context:(void *)context
{
    if(object != myCentralManager)
    {
        return;
    }

    id newDelegate = change[NSKeyValueChangeNewKey];
    id oldDelegate = change[NSKeyValueChangeOldKey];

    if((newDelegate == nil || [newDelegate isKindOfClass:[NSNull class]]) && oldDelegate != self)
    {
        myCentralManager.delegate = self;
    }
}

```

**Comment by qvivr-cris** [comment]
We needed to use CBCentralManagerOptionRestoreIdentifierKey, so we went back to it, just now  identifier we use is now unique each init 

static int identifierIndex = 0;
NSString *restoreIdentifier = [NSString stringWithFormat:@"SWYPCardCentral%i", identifierIndex++];
myCentralManager = [[CBCentralManager alloc] initWithDelegate:self queue:nil options:@{CBCentralManagerOptionRestoreIdentifierKey : restoreIdentifier }];

and we are able to successfully reconnect after a DFU.

**Comment by calvingit** [comment]
@qvivr-cris very good.  My situation is a little different with yours. I have wrapped CBCentralManager with a singleton, and wrapped CBPeripheral as another class named [LGPeripheral](https://github.com/LGBluetooth/LGBluetooth).

**Comment by mostafaberg** [comment]
I see the issue, using a different `CBCentralManager` to reconnect doesn't work all the time, i guess the right thing to do is to restore the `CBCentralManager` from our side after we've "hijacked" it from the developer, this is an enhancement we had planned and should be out in an upcoming version, glad to hear it worked out

**Comment by mostafaberg** [comment]
Possible duplicate https://github.com/NordicSemiconductor/IOS-Pods-DFU-Library/issues/6

**Comment by qvivr-cris** [comment]
It worked once, on one OS (iOS 10.0.2) we will want to test it more.  Yes is there was a process to hand off ownership and take ownership back that would probably be better.  Thank you.

**Comment by mostafaberg** [comment]
Will work on that on the next release and will try to update and will do some tests to verify this works as it should

**Comment by TaiPhamD** [comment]
any work around on restoring the delegate after the DFU class hijack it?

**Comment by mostafaberg** [comment]
@rushstrike There's a good example solution in the comments above [found here](https://github.com/NordicSemiconductor/IOS-Pods-DFU-Library/issues/29#issuecomment-252816614)

**Comment by philips77** [comment]
This will be invalid after #228 gets merged.

**Comment by philips77** [comment]
Fixed in https://github.com/NordicSemiconductor/IOS-Pods-DFU-Library/releases/tag/4.2.0
